### PR TITLE
Implement sns-handling view including a good initial level of tests

### PR DESCRIPTION
**Pull request by risicle**
https://trello.com/c/i30jDD69/14-automatically-virus-scan-uploaded-files

There *probably* should be some more tests around the edge cases here in time, but initially this appears to work quite well.

Example things that are untested:

 - multiple `Records` in a single sns message, though it's unclear whether this actually ever happens with our setup.
 - any one of our calls to external services could in theory fail for random network-related reasons. How many of these cases we should actually cover is up for debate. Currently we test a few of them.

Not included here (though perhaps in time) is handling of infected-file-reversion.

### CODE REVIEW

File: app/authentication.py [hunk 1]
```diff
@@ -17,9 +17,13 @@ def token_is_valid(incoming_token, module):
...
     """Return a list of allowed auth tokens from the application config"""
-    env_variable_name = 'DM_ANTIVIRUS_API_AUTH_TOKENS'
+    env_variable_name = ""
 
-    if module == 'callbacks':
+    if module == "main":
+        env_variable_name = 'DM_ANTIVIRUS_API_AUTH_TOKENS'
+    elif module == 'callbacks':
+        # though SNS uses the "Basic" auth scheme, we can treat the payload section of the header as a single opaque
+        # token set in DM_ANTIVIRUS_API_CALLBACK_AUTH_TOKENS like any normal "Bearer" tokens.
```
Comment (Wynndow) [inline app/authentication.py]: So when SNS uses Basic auth, we're not actually decoding the base64 and splitting it up, like with the Bearer token. We're just taking the base64 encoding as is, and comparing it with our known value from the config?

File: app/callbacks/views/sns.py [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    # check it's for the right topic
+    if topic_name != supported_topic_name:
+        current_app.logger.warning(
+            "Received SubscriptionConfirmation request for unrecognized topic name {topic_name} in {topic_arn}",
+            extra={
+                "topic_name": topic_name,
+                "topic_arn": topic_arn,
+            },
+        )
+        abort(400, f"Unrecognized topic name {topic_name!r}")
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: Is the` __repr__` formatting just to put quotes around it? `topic_name` will only ever be a string, right?

File: app/callbacks/views/sns.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+    # send a request to provided url, ensuring sufficient logging
+    try:
+        with logged_duration(
+            logger=current_app.logger,
+            message=lambda _: (
+                "Made GET request to {target_url} to confirm subscription to topic {topic_arn}"
+                if sys.exc_info()[0] is None else
+                # need to literally format() exception into message as it's difficult to get it injected into extra
+                "Failed to make GET request to {{target_url}} to confirm subscription to {{topic_arn}}: {!r}".format(
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: Are the double curly braces here so that after formatting, the string still has single curly braces around `target_url` and `topic_arn`, which are needed for the log context extra thing?

File: app/callbacks/views/sns.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+                    sys.exc_info()[1]
+                )
+            ),
+            log_level=logging.INFO,
+            condition=True,
+        ) as log_context:
+            topic_arn = body_json["TopicArn"]
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: I think you've already extracted this from the body on line 43, could reuse.

File: app/authentication.py (hunk 1 above)
Comment (risicle) [reply app/authentication.py]: :grin:

File: app/callbacks/views/sns.py (hunk 4 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: Spotted, sir! :mag:

File: app/callbacks/views/sns.py [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+                "topic_arn": topic_arn,
+            })
+            confirmation_response = requests.get(confirmation_url)
+            confirmation_response.raise_for_status()
+    except requests.exceptions.RequestException:
+        abort(400, f"SubscriptionConfirmation request to {confirmation_url!r} failed")
+
+    # check we received what we expected from url
+    subscription_arn = confirmation_request_id = None
+    try:
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: Took me quite a while to work out what was going on here. Might be worth adding more comments?
Also, is there a scenario where subscription_arn is successfully parsed from the xml, but isn't actually an arn? Like if it was `ARN_DOESNT_EXIST`. Or something? Could you check that the retrieved value matches a regex or something?

Comment (Wynndow) [reply app/callbacks/views/sns.py]: Or is it possible that there will be multiple namespaces? If so, would `subscription_arn` be a list? What happens then?

File: app/callbacks/views/sns.py [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+        log_level=logging.INFO,
+        condition=True,
+    ) as log_context:
+        log_context.update({"target_url": url})
+        response = requests.get(url)
+        response.raise_for_status()
+        return response.content
+
+
+# deliberately omitting the optional trailing .cn here
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: What trailing `.cn`? Where is this defined?

File: app/callbacks/views/sns.py [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+VALIDATION_CERTIFICATE_URL_REGEX = r"^https://[-a-z0-9.]+\.amazonaws\.com/"
+# give request extra headroom of a minute here
+VALIDATION_MAX_AGE = datetime.timedelta(hours=1, minutes=1)
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: y?

File: app/callbacks/views/sns.py [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+from dmutils.timing import logged_duration_for_external_request as log_external_request, logged_duration
+
+from app.callbacks import callbacks
+from app.clam import get_clamd_socket
+
+
+@callbacks.route('/')
+@callbacks.route('')
+def callbacks_root():
+    return jsonify(status='ok'), 200
```
[the remark this answers is not in the document]
Comment (Wynndow) [reply app/callbacks/views/sns.py]: Feels redundant as we have a status endpoint

File: app/callbacks/views/sns.py (hunk 6 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: Oh yeah didn't include the context - this basically like the default regex from the `validatesns` module but doesn't allow the optional `.cn` which it has.

File: app/callbacks/views/sns.py [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+        )
+    except (validatesns.ValidationError, requests.exceptions.RequestException) as e:
+        current_app.logger.warning("SNS request body failed signature validation: {validation_error}", extra={
+            "validation_error": e,
+        })
+        abort(400, "SNS request body failed signature validation")
+
+    supported_topic_name = f"s3_file_upload_notification_{current_app.config['DM_ENVIRONMENT']}"
+
+    if body_dict["Type"] == "SubscriptionConfirmation":
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: Any chance of a key error here? Will there definitely be a 'Type' key?

File: app/callbacks/views/sns.py (hunk 5 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: I think ARNs are kinda supposed to be opaque (even though they're not totally) and so I'm just treating it like a string. If they give us a value that isn't an ARN, ... well, ... they shouldn't. It would be against spec. We can't guard against aws breaking all its own rules. I mean whatever we get there kinda *is* our subscription ARN by definition.

The only thing that's going on with the namespaces is I'm trying to get it to leave me alone about namespaces. I'm not doing the `n:` stuff for fun - because the xml has an `xmlns=` attr at the top level, lxml will want me to be explicit about all the namespaces in the xpath expression. Assigning whatever the toplevel namespace uri is to `n` is just my way of making that nonsense as brief as possible.

File: app/callbacks/views/sns.py [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+            }
+            log_context.update(base_log_context)
+
+            # TODO abort if file too big?
+
+            s3_client = boto3.client("s3", region_name=record["awsRegion"])
+
+            with log_external_request(
+                "S3",
+                "get object tagging [{s3_bucket_name}/{s3_object_key} versionId {s3_object_version}]",
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: Does this need to be an f string?

File: app/callbacks/views/sns.py (hunk 5 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: On the other hand, perhaps I shouldn't be lenient, and simply hard code the expected namespace uri and allow it to fail if they ever change that...? There is a degree to which I'm taking the battery out of the smoke alarm here.

File: app/callbacks/views/sns.py (hunk 7 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: Just using my :crystal_ball: to predict obvious unnecessary failures that might occur and annoy us.

File: app/callbacks/views/sns.py (hunk 9 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: :man_shrugging: if there ain't, maybe we *should* error and know about it?

File: app/callbacks/views/sns.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (risicle) [reply app/callbacks/views/sns.py]: So does the api. Still can't figure out why it has the equivalent of this though.

File: app/callbacks/views/sns.py (hunk 7 above)
Comment (Wynndow) [reply app/callbacks/views/sns.py]: It's gone from 1 hour, to 1 hour and 1 minute?

Comment (risicle) [reply app/callbacks/views/sns.py]: :grin:

File: app/callbacks/views/sns.py [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+
+            tagging_tag_set = _tag_set_updated_with_value(tagging_tag_set, "avStatus", new_av_status_json)
+
+            with log_external_request(
+                "S3",
+                "put object tagging [{s3_bucket_name}/{s3_object_key} versionId {s3_object_version}]",
+                logger=current_app.logger,
+            ) as log_context_s3:
+                log_context_s3.update(base_log_context)
+                tagging_tag_set = s3_client.put_object_tagging(
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: I don't think you need to reassign `tagging_tag_set` here - it's not used again.

File: app/callbacks/views/sns.py [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+                # TODO?     if there is no such version:
+                # TODO?         S3: upload fail whale?
+                # TODO?     else copy that version to become new "current" ver for this key, ensuring to copy its tags
+                # TODO?         note the impossibility of doing this without some race conditions
+
+                notify_client = DMNotifyClient(current_app.config["DM_NOTIFY_API_KEY"])
+                try:
+                    notify_client.send_email(
+                        current_app.config["DM_DEVELOPER_VIRUS_ALERT_EMAIL"],
+                        template_id=current_app.config["NOTIFY_TEMPLATES"]["developer_virus_alert"],
```
Comment (Wynndow) [inline app/callbacks/views/sns.py]: This template is very.....you.

File: app/callbacks/views/sns.py (hunk 9 above)
Comment (Wynndow) [reply app/callbacks/views/sns.py]: Yep, fair enough

File: app/callbacks/views/sns.py (hunk 6 above)
Comment (Wynndow) [reply app/callbacks/views/sns.py]: The url can have a trailing `.cn`? Why are we disallowing it?

File: app/callbacks/views/sns.py (hunk 5 above)
Comment (Wynndow) [reply app/callbacks/views/sns.py]: I think just stick that comment above about namespaces in.

File: app/callbacks/views/sns.py (hunk 12 above)
Comment (risicle) [reply app/callbacks/views/sns.py]: Isn't it just?

**Comment by risicle** [comment]
Changes pushed, @benvand can I merge this?
